text/3389-manifest-lint.md

```diff
@@ -0,0 +1,398 @@
+- Feature Name: `manifest-lint`
+- Start Date: 2023-02-14
+- RFC PR: [rust-lang/rfcs#3389](https://github.com/rust-lang/rfcs/pull/3389)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add a `[lints]` table to `Cargo.toml` to configure reporting levels for
+rustc and other tool lints.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, you can configure lints through
+- `#[<level>(<lint>)]` or `#![<level>(<lint>)]`, like `#[forbid(unsafe)]`
+  - But this doesn't scale up with additional targets (benches, examples,
+    tests) or workspaces
+- On the command line, like `cargo clippy -- --forbid unsafe`
+  - This puts the burden on the caller
+- Through `RUSTFLAGS`, like `RUSTFLAGS=--forbid=unsafe cargo clippy`
+  - This puts the burden on the caller
+- In `.cargo/config.toml`'s `target.*.rustflags`
+  - This couples you to the running in specific directories and not running in
+    the right directory causes rebuilds
+  - The cargo team has previously stated that
+    [they would like to see package-specific config moved to manifests](https://internals.rust-lang.org/t/proposal-move-some-cargo-config-settings-to-cargo-toml/13336/14?u=epage)
+
+We would like a solution that makes it easier to share across targets and
+packages for all tools.
+
+See also
+- [rust-lang/rust-clippy#1313](https://github.com/rust-lang/rust-clippy/issues/1313)
+- [rust-lang/cargo#5034](https://github.com/rust-lang/cargo/issues/5034)
+- [EmbarkStudios/rust-ecosystem#59](https://github.com/EmbarkStudios/rust-ecosystem/issues/59)
+- [Proposal: Cargo Lint configuration](https://internals.rust-lang.org/t/proposal-cargo-lint-configuration/9135)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A new `lints` table would be added to configure lints:
+```toml
+[lints.rust]
+unsafe = "forbid"
```

Comment:
This is covered in the alternatives section

> Instead of the `[lints]` table being `lint = "level"`, we could organize
it around `level = ["lint", ...]` like some other linters do (like
[ruff](https://beta.ruff.rs/docs/configuration/)) but this works better for
logically organizing lints, highlighting what changed in diffs, and for
possibly adding lint-specific configuration in the future.


> Also other tools (like cargo-cranky) already use this syntax, and going through the cargo issue about this feature the level = ["lint1", "lint2", ...] is what most people expect it to be.

I explored this a bit in the prior art section.  Python tends to use `level = [lint]`.  golangci-lint does the same,.  Javascript's eslint and Ruby's rubocop are closer to what this RFC references.   Which people expect is likely related to what ecosystem they come from.  Even that isn't always a sure bet.  Most of my exposure to linters is with Python and the `level = [lint]` model always bothered me.

Reply:
> Most of my exposure to linters is with Python and the `level = [lint]` model always bothered me.

Most of my exposure with lint configuration is also with `python`, and I come to like the `level = [lint]` style, because I find it easier to write and read a list of lints under a category, rather then go through each lint to know which one is allowed and which one is denied.

Why does this way of writing it bothers you?

I guess this is really highly dependent on personal preference...

Reply:
I also think that this is more egonomic:
```text
[lints]
<level>=[
  "<rules>",
  "<rules>",
  ...
]
```

concrete example:
```toml
[lints]
allow = [
  "dead_code", 
  "unused",
  "clippy::module_name_repetitions",
]
warn = [
  "clippy::disallowed-names",
]
```

Reply:
One input (not requirement) for the lint format is whether we leave the door open for lint configuration (as mentioned in that block I quoted above).  At least to me, lint configuration with the `level = lint` organization is more awkward.

Reply:
I would say that lint configuration is an edge case. And the only application is `priority` or potentially `rust-version`. This could also be modeled with

```toml
[lints.rust]
allow = [
  { name="dead_code", priority=1 }, 
]
```

instead of 

```toml
[lints.rust]
dead_code = { level="allow", priority=1 }
```

Reply:
We discussed this in the cargo team meeting.  While I might not be capturing all the views of the attendees, the thoughts were:
- One element of this conversation is what people are used to
- For `level = [lint]`
  - Less verbose
- For `lint = level`
  - Some felt that they were more likely to think "lint" and look that up, rather than "level" then "lint".
  - You don't get lost in [large lint lists](https://github.com/EmbarkStudios/rust-ecosystem/blob/main/lints.toml) like people do in dependency tables today
  - Depending on the users mental model, if we do add overriding of inherited lint configuration, the package's table being merged into the workspace table conceptually makes more sense (and is easier to implement correctly)
  - The potential for lint configuration as mentioned

Overall, we are leaning towards `lint = level` but the priority conversation could change that.

Reply:
> You don't get lost in [large lint lists](https://github.com/EmbarkStudios/rust-ecosystem/blob/main/lints.toml) like people do in dependency tables today

IMO `level = [lint]` actually fares better in this sort of test, `lint = level` has two major downsides to scannability: the jagged ` = "warn"` edge makes picking out the lint names harder, and the ability to mix levels makes it easy to mis-identify what level a specific lint is at.

```
warn = [
  "clippy::all",
  "clippy::await_holding_lock",
  "clippy::char_lit_as_u8",
  "clippy::checked_conversions",
  "rust_2018_idioms",
]
deny = [
  "unsafe_code",
]
```
```
"clippy::all" = "warn"
"clippy::await_holding_lock" = "warn"
"clippy::char_lit_as_u8" = "warn"
"clippy::checked_conversions" = "warn"
"unsafe_code" = "deny"
"rust_2018_idioms" = "warn"
```

Reply:
> * Some felt that they were more likely to think "lint" and look that up, rather than "level" then "lint".

I'm not sure, I understand what this is supposed to mean.

> * You don't get lost in [large lint lists](https://github.com/EmbarkStudios/rust-ecosystem/blob/main/lints.toml) like people do in dependency tables today

I also don't understand this point and think it's the other way around. I don't see how one would get less lost with the `lint = "level"` syntax:

<details>
<summary>Linked example with both syntaxes</summary>

```toml
[lints.rust]
unsafe_code = "deny"
future_incompatible = "warn"
nonstandard_style = "warn"
rust_2018_idioms = "warn"

[lints.clippy]
all = "warn"
await_holding_lock = "warn"
char_lit_as_u8 = "warn"
checked_conversions = "warn"
dbg_macro = "warn"
debug_assert_with_mut_call = "warn"
doc_markdown = "warn"
empty_enum = "warn"
enum_glob_use = "warn"
exit = "warn"
expl_impl_clone_on_copy = "warn"
explicit_deref_methods = "warn"
explicit_into_iter_loop = "warn"
fallible_impl_from = "warn"
filter_map_next = "warn"
flat_map_option = "warn"
float_cmp_const = "warn"
fn_params_excessive_bools = "warn"
from_iter_instead_of_collect = "warn"
if_let_mutex = "warn"
implicit_clone = "warn"
imprecise_flops = "warn"
inefficient_to_string = "warn"
invalid_upcast_comparisons = "warn"
large_digit_groups = "warn"
large_stack_arrays = "deny"
large_types_passed_by_value = "warn"
let_unit_value = "warn"
linkedlist = "warn"
lossy_float_literal = "warn"
macro_use_imports = "warn"
manual_ok_or = "warn"
map_err_ignore = "warn"
map_flatten = "warn"
map_unwrap_or = "warn"
match_on_vec_items = "warn"
match_same_arms = "warn"
match_wild_err_arm = "warn"
match_wildcard_for_single_variants = "warn"
mem_forget = "warn"
mismatched_target_os = "warn"
missing_enforced_import_renames = "warn"
mut_mut = "warn"
mutex_integer = "warn"
needless_borrow = "warn"
needless_continue = "warn"
needless_for_each = "warn"
option_option = "warn"
path_buf_push_overwrite = "warn"
ptr_as_ptr = "warn"
rc_mutex = "warn"
ref_option_ref = "warn"
rest_pat_in_fully_bound_structs = "warn"
same_functions_in_if_condition = "warn"
semicolon_if_nothing_returned = "warn"
single_match_else = "warn"
string_add_assign = "warn"
string_add = "warn"
string_lit_as_bytes = "warn"
string_to_string = "warn"
todo = "warn"
trait_duplication_in_bounds = "warn"
unimplemented = "warn"
unnested_or_patterns = "warn"
unused_self = "warn"
useless_transmute = "warn"
verbose_file_reads = "warn"
zero_sized_map_values = "warn"
```

As an exercise for the reader, I've hidden a "deny" in the list. Can you spot it? 

---

```toml
[lints.rust]
deny = [
    "unsafe_code",
]
warn = [
    "future_incompatible",
    "nonstandard_style",
    "rust_2018_idioms",
]

[lints.clippy]
warn = [
    "all",
    "await_holding_lock",
    "char_lit_as_u8",
    "checked_conversions",
    "dbg_macro",
    "debug_assert_with_mut_call",
    "doc_markdown",
    "empty_enum",
    "enum_glob_use",
    "exit",
    "expl_impl_clone_on_copy",
    "explicit_deref_methods",
    "explicit_into_iter_loop",
    "fallible_impl_from",
    "filter_map_next",
    "flat_map_option",
    "float_cmp_const",
    "fn_params_excessive_bools",
    "from_iter_instead_of_collect",
    "if_let_mutex",
    "implicit_clone",
    "imprecise_flops",
    "inefficient_to_string",
    "invalid_upcast_comparisons",
    "large_digit_groups",
    "large_stack_arrays",
    "large_types_passed_by_value",
    "let_unit_value",
    "linkedlist",
    "lossy_float_literal",
    "macro_use_imports",
    "manual_ok_or",
    "map_err_ignore",
    "map_flatten",
    "map_unwrap_or",
    "match_on_vec_items",
    "match_same_arms",
    "match_wild_err_arm",
    "match_wildcard_for_single_variants",
    "mem_forget",
    "mismatched_target_os",
    "missing_enforced_import_renames",
    "mut_mut",
    "mutex_integer",
    "needless_borrow",
    "needless_continue",
    "needless_for_each",
    "option_option",
    "path_buf_push_overwrite",
    "ptr_as_ptr",
    "rc_mutex",
    "ref_option_ref",
    "rest_pat_in_fully_bound_structs",
    "same_functions_in_if_condition",
    "semicolon_if_nothing_returned",
    "single_match_else",
    "string_add_assign",
    "string_add",
    "string_lit_as_bytes",
    "string_to_string",
    "todo",
    "trait_duplication_in_bounds",
    "unimplemented",
    "unnested_or_patterns",
    "unused_self",
    "useless_transmute",
    "verbose_file_reads",
    "zero_sized_map_values",
]
```
</details>

> * Depending on the users mental model, if we do add overriding of inherited lint configuration, the package's table being merged into the workspace table conceptually makes more sense (and is easier to implement correctly)

I wouldn't put a lot of thought behind what's easier to implement in `cargo`, but rather what's easier to use.

Merging `warn` tables vs merging tables of lint are not two very different mental models and both are equally easy to understand IMO.

> * The potential for lint configuration as mentioned

That is still possible by allowing tables in the `warn`/`allow`/... lists.

---

Another argument for the `level = [lint]` syntax is, that it potentially doesn't require splitting up the lint tables by tool. See https://github.com/rust-lang/rfcs/pull/3389#discussion_r1153145977. But this might still be necessary, because `rustdoc` is an outlier  IIUC.

Reply:
AFAIK `rustdoc` isn't an outlier, it also ignores all unknown tool lints (`cargo rustdoc -- -W clippy::pedantic` runs just fine).

Reply:
> > You don't get lost in [large lint lists](https://github.com/EmbarkStudios/rust-ecosystem/blob/main/lints.toml) like people do in dependency tables today
>
> IMO level = [lint] actually fares better in this sort of test, lint = level has two major downsides to scannability: the jagged  = "warn" edge makes picking out the lint names harder, and the ability to mix levels makes it easy to mis-identify what level a specific lint is at.

When looking at a specific lint, you know what its level is vs being in a long list of lints and losing track of what level you are in.

What you are mentioning is something different, quickly scanning for lint levels.

> I wouldn't put a lot of thought behind what's easier to implement in cargo, but rather what's easier to use.

To be clear, I wasn't.  I emphasized the user workflow and parenthetically talked about the implementation.

> > The potential for lint configuration as mentioned
>
> That is still possible by allowing tables in the warn/allow/... lists. See https://github.com/rust-lang/rfcs/pull/3389#discussion_r1155824507

All things being (mostly) equal, we aren't too big of fans of complex/deep nested inline data structures.  We only considered this syntax in an alternative to `priority` because our concerns for `priority` helped shift the balance enough.

Reply:
> When looking at a specific lint, you know what its level is vs being in a long list of lints and losing track of what level you are in.

I would expect the question "What lints are set to warn/deny?" come up more often, than "What level does this exact lint have?". And the second question is also easily answerable by just jumping to the start of the list with the `level = [lint]` syntax. The first question is really hard to answer with the `lint = level` syntax, if the lints are not sorted by level (which the other syntax does).

> To be clear, I wasn't. I emphasized the user workflow

ACK. I still don't see a benefit of either syntax for the user regarding packages/workspaces.

> All things being (mostly) equal, we aren't too big of fans of complex/deep nested inline data structures

I agree that the less nesting of the `lint = level` syntax is nicer for this use case. However, I don't think this is the main use case - far from it. Most lints won't need any configuration or `priority`. So I don't think this outweighs the benefits of the `level = [lint]` syntax.

Reply:
I've updated the RFC to include some of the discussion from here on which schema to use